api: listVM API improvement followup, change returning of stats detail

rohityadavcloud opened the pull request:
### Description

Followup from https://github.com/apache/cloudstack/pull/8782

- Changes behaviour of details param handling via global setting:
  - listVirtualMachines API: when the details param is not provided, it returns whether stats are returned controlled by a new global setting `list.vm.default.details.stats`
  - listVirtualMachinesMetrics API: when the details param is not provided, it uses `all` details including `stats`
- Users who are affected slow performance of the listVirtualMachines API response time can set `list.vm.default.details.stats` to `false`
- Remove ConfigKey vm.stats.increment.metrics.in.memory which was renamed to `vm.stats.increment.metrics` in #5984 and also remove unused/unnecessary global settings via upgrade path
- Changes default value of VM stats accumulation setting `vm.stats.increment.metrics` to false until a better solution emerges. Since #5984, this is true and during the execution of listVM APIs the stats are clubbed/calculated which can immensely slow down list VM API calls. Any costly operations such as summing of stats shouldn't be done during the course of a synchronous API, such as the list VM API.
- Fix UI that uses listVirtualMachinesMetrics to not call `stats` detail when in list view without metrics selected.

### Types of changes

- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [x] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI

### Feature/Enhancement Scale or Bug Severity

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [x] Major
- [ ] Minor
- [ ] Trivial

blueorangutan commented:
@rohityadavcloud a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: ✔️ el7 ✔️ el8 ✔️ el9 ✔️ debian ✔️ suse15. SL-JID 9796

rohityadavcloud commented:
@blueorangutan test

blueorangutan commented:
@rohityadavcloud a [SL] Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

DaanHoogland reviewed (approved):
clgtm

blueorangutan commented:
<b>[SF] Trillian test result (tid-10375)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 49031 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr9177-t10375-kvm-centos7.zip
Smoke tests completed. 131 look OK, 0 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---

GutoVeronezi commented:
-1 on changing the `list.vm.default.details.stats` default value. Refer to the discussion in #8782.

rohityadavcloud commented:
@GutoVeronezi this isn't a vote, could you explain why you're opposing now that the default value change when we've agreed in #8782 to address this via documentation of the change in release notes?

DaanHoogland commented:
@GutoVeronezi , I understand your point, but we do not have a simple usable procedure yet. The implicit procedure was "we don't introduce any backwards incompatability, .. unless we all agree". And so in this case you are right. That does not remove the need to decide on individual issues.
Maybe @rohityadavcloud is a bit quick art the gun, but it makes sense to first introduce a setting and than later reverse the default. We have done it like that in the past.  Maybe we can add this to #8970 (if it is not included yet).

So how do you propose we take this forward? cc @JoaoJandre @sureshanaparti @weizhouapache

GutoVeronezi commented:
> @GutoVeronezi , I understand your point, but we do not have a simple usable procedure yet. The implicit procedure was "we don't introduce any backwards incompatability, .. unless we all agree". And so in this case you are right. That does not remove the need to decide on individual issues. Maybe @rohityadavcloud is a bit quick art the gun, but it makes sense to first introduce a setting and than later reverse the default. We have done it like that in the past. Maybe we can add this to #8970 (if it is not included yet).
> 
> So how do you propose we take this forward? cc @JoaoJandre @sureshanaparti @weizhouapache

As we already have a configuration to change the behavior, we should keep the current behavior and guide people through the release notes on how to change it. Then, we should continue the #8970 discussion. Once a process is defined and documented, we can look back to the configuration and change the default according to the policy we defined (or even remove the configuration if we decide so). Other systems do that way, like GitLab and OpenStack; the problem is that we do not have this defined and documented yet.

rohityadavcloud commented:
@GutoVeronezi by your own admission I changed my stand, closed my PR to support Joao's, and built support and consensus with others on Joao's PR and merge his PR. I also don't appreciate some of your remarks and behaviour towards me. It's unfair to me when I'm genuinely trying to help and support everybody and anybody in the community.

I don't see anything wrong with somebody changing their stand when presented with new information or convincing arguments. My reason for changing my stand wrt the default behaviour is that most users don't consume stats (outside of UI) via the list APIs, but many integrations such as Terraform provider, k8s-provider, CAPC, users ansible scripts etc (point of note that these integrations can't be changed) call the list VM APIs and can benefit from any potential speed up. @GutoVeronezi @JoaoJandre and anyone else - what are your reasons of changing your stand wrt the default behaviour now?

For example, look at the two user reported issues that also contributed to change my stand:
https://github.com/apache/cloudstack/issues/8975
https://github.com/apache/cloudstack/issues/7910

I think we introduced the behaviour, and we can discuss and agree if we want to change that if that is largely beneficial to users. I don't think it would detrimental for large number of users if they have gain speedups and still had a global setting to change the behaviour.

That said, I don't have the energy or bandwidth to fight over this, if this is the only thing blocking this I can change the global setting. Have a look, review and advise - @JoaoJandre @DaanHoogland @weizhouapache @sureshanaparti @GutoVeronezi @shwstppr and others...

shwstppr reviewed (approved):
Code LGTM

- Slow VM listing needs to be addressed as it has already been reported in various forms
- I don't see changing the value for a boolean global config as a big breaking change if we plan to document it properly. Users can revert it after upgrade if they wish to continue using incremental stats.
- If we did not agree on changing config value in the past we can discuss it now with pros and cons
- I don't think it makes sense to block this change till we decide on #8970. It deals with wider discussion and there are no clear outcomes in the past 45 days or so

GutoVeronezi commented:
@rohityadavcloud, I did not judge your change of opinion. If you read it carefully, you will see that what I am doing is pointing out that there was no consensus on the change of the default behavior in the whole discussion. There were no direct attacks on anyone, I am just making the timeline clear. 

We would like to introduce some breaking changes and changes to default behavior, but we do not have a documented process for this, and doing it on a case-by-case basis is prone to heated discussions and unfair decisions. Discussion #8970 is there to make these processes well-defined and documented so we can avoid this kind of discussions in the future.

> * I don't think it makes sense to block this change till we decide on [Define a release schedule for the project #8970](https://github.com/apache/cloudstack/discussions/8970). It deals with wider discussion and there are no clear outcomes in the past 45 days or so

@shwstppr 

If more people interact there, we can get some outcomes :smiley:

rohityadavcloud commented:
Amusing, we're going in circles now and obstinately determined about it.

I've tried my part, got non-technical and discourteous reasons to block the PR, coerced to unrelated agenda. I'm also flattered some want my attention and validation so bad.

I'm gonna sit on the fence on this one, for now; save my time and energy.

`list.vm.default.details.stats` defaults to true now.

@blueorangutan package

blueorangutan commented:
@rohityadavcloud a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

rohityadavcloud commented:
@blueorangutan ui

blueorangutan commented:
@rohityadavcloud a Jenkins job has been kicked to build UI QA env. I'll keep you posted as I make progress.

blueorangutan commented:
UI build: :heavy_check_mark:
Live QA URL: https://qa.cloudstack.cloud/simulator/pr/9177 (QA-JID-373)

rohityadavcloud commented:
Requesting re-review as code has changed a bit

@blueorangutan test

blueorangutan commented:
@rohityadavcloud a [SL] Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>[SF] Trillian Build Failed (tid-10429)<b/>

rohityadavcloud commented:
@blueorangutan test

blueorangutan commented:
@rohityadavcloud a [SL] Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

shwstppr reviewed (approved):
Code LGTM

blueorangutan commented:
<b>[SF] Trillian test result (tid-10430)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 43903 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr9177-t10430-kvm-centos7.zip
Smoke tests completed. 130 look OK, 1 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_02_trigger_shutdown | `Failure` | 341.76 | test_safe_shutdown.py
